[Windows] Force reinstall if configuration changed [AP-1946]

chouquette opened the pull request:
This PR adds support for optional features exposed by the MSI installer.
It will probe the available features and will trigger a reinstall if an optional features must be enabled/disabled (the installer doesn't support in place modification of the installed features)

While this is of limited interest given that the installer always installs all features since 7.46, this paves the way for future optional features that we might not want to systematically install

KevinFairise2 commented on tasks/pkg-windows.yml:
```diff
@@ -4,6 +4,31 @@
+  set_fact:
+    agent_datadog_skip_install: >-
+      {{ agent_config_install_args.exists and
+      agent_config_install_args.value != agent_win_install_args }}
```
Shouldn't we skip the install when the provided args match the previous one stored in the registry? It looks like `agent_config_install_args.value != agent_win_install_args` will set `agent_datadog_skip_install` to true when they are not the same.

chouquette replied:
Hm, indeed you are correct the logic is inverted, which means something else is wrong since the install was still forced during my tests

chouquette replied:
I found the issue, it was somewhere between the chair and the keyboard (I forgot to pin the version so I was checking with a `agent_datadog_skip_install` set to false, meaning I didn't test anything valuable)

KevinFairise2 commented on tasks/pkg-windows.yml:
```diff
@@ -67,23 +84,24 @@
   register: agent_datadog_agent_install
   when: (not agent_datadog_skip_install) and (not ansible_check_mode)
 
+- name: Store successful instal args
```
```suggestion
- name: Store successful install args
```

chouquette committed:
tasks: windows: add a task to probe installed features

This will help for [AP-1946] and allows us to detect if the NPM driver
is installed on old agent installations

chouquette committed:
tasks: windows: fix typo

chouquette committed:
tasks: windows: include pkg-windows-opts earlier

We will soon override agent_datadog_skip_install within this file so the
updated value must be available ASAP

chouquette committed:
tasks: windows: allow the agent to be uninstalled

chouquette committed:
tasks: windows: reinstall the agent if optional features changed

Fix [AP-1946]

chouquette committed:
fix ansible lint errors

chouquette committed:
Apply review suggestions

chouquette commented on tasks/pkg-windows-opts.yml:
```diff
@@ -12,6 +16,23 @@
     agent_win_install_args: "{{ win_install_args }} DDAGENTUSER_NAME={{ datadog_windows_ddagentuser_name }}"
   when: datadog_windows_ddagentuser_name | default('', true) | length > 0
 
+- name: Get Windows Agent config
```
renamed

chouquette commented on tasks/pkg-windows-opts.yml:
```diff
@@ -90,3 +111,16 @@
   when: agent_datadog_sysprobe_enabled
+
+# Check for potential config changes that require a reinstall:
+- name: Check NPM config change
```
renamed

chouquette commented on tasks/pkg-windows-opts.yml:
```diff
@@ -90,3 +111,16 @@
+  # reported as such regardless of the provided parameters. Since we don't
+  # want to force the install for nothing in those version, we skip the step
+  # when the agent is 7.45+
+  when: >-
```
Added a debug message to let the users know

chouquette commented:
Thanks for the review! I pushed an additional commit per your suggestions (and rebased on top of main)

KevinFairise2 reviewed (approved):
LGTM!
